framework/src/araui/core/ui_core.c

```diff
@@ -0,0 +1,837 @@
+#endif
+
+	if (ui_request_callback_init() != UI_OK) {
+		ui_hal_deinit();
+		ui_view_list_deinit();
+		ui_view_update_list_deinit();
+		UI_LOGE("Error: UI_INIT_FAILURE.\n");
+		return UI_INIT_FAILURE;
+	}
+
+	g_core.state = UI_CORE_STATE_RUNNING;
+
+	if (pthread_create(&g_core.pid, &attr, _ui_core_thread_loop, NULL)) {
+		ui_hal_deinit();
+		ui_view_list_deinit();
+		ui_view_update_list_deinit();
+		ui_request_callback_deinit();
+		g_core.state = UI_CORE_STATE_STOP;
+		UI_LOGE("Error: UI_INIT_FAILURE.\n");
+		return UI_INIT_FAILURE;
+	}
+
+#if !defined(UI_PLATFORM_DARWIN)
+	if (pthread_setname_np(g_core.pid, UI_CORE_THREAD_NAME)) {
+		UI_LOGE("error: failed to set pthread name!\n");
+	}
+#endif
+
+	g_core.caller_pid = getpid();
+
+	UI_LOGD("UI_OK.\n");
+	
+	return UI_OK;
+}
+
+ui_error_t ui_stop(void)
+{
+	if (!ui_is_running()) {
+		UI_LOGE("Error: UI_NOT_RUNNING.\n");
+		return UI_NOT_RUNNING;
+	}
+
+	if (g_core.caller_pid != getpid()) {
+		UI_LOGE("Error: Permission denied.\n");
+		return UI_OPERATION_FAIL;
+	}
+
+	g_core.state = UI_CORE_STATE_STOPPING;
+
+	if (pthread_join(g_core.pid, NULL) != OK) {
+		UI_LOGE("pthread_join failed.\n");
+		return UI_OPERATION_FAIL;
+	}
+
+	if (ui_request_callback_deinit() != UI_OK) {
+		UI_LOGE("ui_request_callback_deinit failed.\n");
+		return UI_OPERATION_FAIL;
+	}
+
+	if (ui_view_update_list_deinit() != UI_OK) {
+		UI_LOGE("ui_view_update_list_deinit failed.\n");
+		return UI_OPERATION_FAIL;
+	}
+
+	if (ui_view_list_deinit() != UI_OK) {
+		UI_LOGE("ui_view_list_deinit failed.\n");
+		return UI_OPERATION_FAIL;
+	}
+
+	UI_LOGD("UI_OK.\n");
+
+	return UI_OK;
+}
+
+static ui_error_t _ui_process_widget_recur(ui_widget_body_t *widget, uint32_t dt)
+{
+	ui_list_node_t *node;
+
+	if (!widget) {
+		UI_LOGE("Error: widget is null!\n");
+		return UI_INVALID_PARAM;
+	}
+
+	if (widget->tween_cb) {
+		if (widget->tween_info.t > widget->tween_info.d) {
+			widget->tween_info.t = widget->tween_info.d;
+			widget->tween_cb((ui_widget_t)widget, widget->tween_info.t);
+			widget->tween_cb = UI_NULL;
+
+			if (ui_request_callback(_ui_call_tween_finished_cb, widget) != UI_OK) {
+				UI_LOGE("Error: cannot make a request to call.tween_finished_cb!\n");
+			}
+		} else {
+			widget->tween_cb((ui_widget_t)widget, widget->tween_info.t);
+			widget->tween_info.t += dt;
+		}
+	}
+
+	if (widget->tick_cb) {
+		widget->tick_cb((ui_widget_t)widget, dt);
+	}
+
+	if (widget->interval_cb) {
+		widget->interval_info.current += dt;
+		if (widget->interval_info.current >= widget->interval_info.timeout) {
+			widget->interval_info.current -= widget->interval_info.timeout;
+			widget->interval_cb((ui_widget_t)widget);
+		}
+	}
+
+	if (widget->visible) {
+		if (widget->update_cb) {
+			widget->update_cb((ui_widget_t)widget, dt);
+		}
+
+		node = ui_list_loop_begin(&widget->children);
+		while (node != ui_list_loop_end(&widget->children)) {
+			ui_widget_body_t *child = (ui_widget_body_t *)node->data;
+			_ui_process_widget_recur(child, dt);
```

Comment:
child is always valid here? actually this function almost always return UI_OK but return error when widget is null.
so, if child is always valid then it is very stable but...

Reply:
With our policy,

- Iteration linked list must be run in a thread. (Isn't run in multithread)
- The item that inserted to the list is guaranteed.